internal/slsapi/slsapi.go

```diff
@@ -0,0 +1,190 @@
+	"github.com/sirupsen/logrus"
+)
+
+const DefaultSlsUrl string = "http://cray-sls/"
+
+type SLS struct {
+	Url    *url.URL
+	Client *retryablehttp.Client
+}
+
+type slsReady struct {
+	Ready  bool   `json:"Ready"`
+	Reason string `json:"Reason,omitempty"`
+	Code   int    `json:"Code,omitempty"`
+}
+
+type ManagementSwitch struct {
+	Parent          string                        `json:"Parent"`
+	Children        []string                      `json:"Children,omitempty"`
+	Xname           string                        `json:"Xname"`
+	Type            sls_common.HMSStringType      `json:"Type"`
+	Class           sls_common.CabinetType        `json:"Class"`
+	TypeString      base.HMSType                  `json:"TypeString"`
+	LastUpdated     int64                         `json:"LastUpdated,omitempty"`
+	LastUpdatedTime string                        `json:"LastUpdatedTime,omitempty"`
+	ExtraProperties sls_common.ComptypeMgmtSwitch `json:"ExtraProperties,omitempty"`
+}
+
+var serviceName string
+
+// Allocate and initialize new SLS struct.
+func NewSLS(slsUrl string, httpClient *retryablehttp.Client, svcName string) *SLS {
+	var err error
+	serviceName = svcName
+	sls := new(SLS)
+	if sls.Url, err = url.Parse(slsUrl); err != nil {
+		sls.Url, _ = url.Parse(DefaultSlsUrl)
+	} else {
+		// Default to using http if not specified
+		if len(sls.Url.Scheme) == 0 {
+			sls.Url.Scheme = "http"
+		}
+	}
+
+	// Create an httpClient if one was not given
+	if httpClient == nil {
+		sls.Client = retryablehttp.NewClient()
+		sls.Client.HTTPClient.Transport = &http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}
+		sls.Client.RetryMax = 5
+		sls.Client.HTTPClient.Timeout = time.Second * 40
+		//turn off the http client loggin!
+		tmpLogger := logrus.New()
+		tmpLogger.SetLevel(logrus.PanicLevel)
+		sls.Client.Logger = tmpLogger
+	} else {
+		sls.Client = httpClient
+	}
+	return sls
+}
+
+// Ping SLS to see if it is ready.
+func (sls *SLS) IsReady() (bool, error) {
+	var ready slsReady
+	if sls.Url == nil {
+		return false, fmt.Errorf("SLS struct has no URL")
+	}
+	uri := sls.Url.String() + "/ready"
+	req, err := http.NewRequest(http.MethodGet, uri, nil)
+	if err != nil {
+		return false, err
+	}
+	body, err := sls.doRequest(req)
+	if err != nil {
+		return false, err
+	}
+
+	err = json.Unmarshal(body, &ready)
+	if err != nil {
+		return false, err
+	}
+
+	if !ready.Ready {
+		err = fmt.Errorf("%d - %s", ready.Code, ready.Reason)
+	}
+	return ready.Ready, err
+}
+
+// Query SLS for all of the management switches it has.
+func (sls *SLS) GetAllSwitchInfo() (switchHardware []ManagementSwitch, err error) {
+	// Validate inputs
+	if sls.Url == nil {
+		err = fmt.Errorf("SLS struct has no URL")
+		return
+	}
+
+	// Construct a GET to '/search/hardware?type=comptype_mgmt_switch'
+	// for SLS to get all of the management switches in SLS.
+	uri := sls.Url.String() + "/search/hardware?type=" + sls_common.MgmtSwitch.String()
```

Comment:
This will change to include the other switch types.

The other switches may not have the same extra properties but they won't have any more than are defined in the sls_common.ComptypeMgmtSwitch struct. They'll Unmarshal() just fine and I'll make changes to handle missing fields for the other switches.